### PR TITLE
Feat (Withdraw): Add its API

### PR DESCRIPTION
**Pull request by OjusWiZard**
## Proposed changes

Add `POST /api/services/{service}/onchain/withdraw` for withdrawing all the funds to a wallet

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: operate/cli.py [hunk 1]
```diff
@@ -663,6 +666,89 @@ async def _stop_service_onchain(request: Request) -> JSONResponse:
...
+                hash=request.path_params["service"],
+                withdrawal_address=withdrawal_address
+            )
+
+            # drain OLAS from the master safe
+            chain_config = service.chain_configs[service.home_chain_id]
+            master_wallet: EthereumMasterWallet = service_manager.wallet_manager.load(
+                ledger_type=chain_config.ledger_config.type
+            )
+            ocm = service_manager.get_on_chain_manager(ledger_config=chain_config.ledger_config)
```
Comment (jmoreira-valory) [inline operate/cli.py:705]: On-chain manager is used for EVM operations without Safe (legacy code). This should not be used, in favour of EthSafeTxBuilder

Comment (jmoreira-valory) [reply operate/cli.py:705]: All OnChainManager code will be removed, as we are not considering use cases without Safe. cc @oaksprout

**Review by jmoreira-valory — CHANGES_REQUESTED** [review]
Refactor usage of ocm.

File: operate/cli.py (hunk 1 above)
Comment (0xArdi) [reply operate/cli.py:705]: I agree with @jmoreira-valory, lets use `EthSafeTxBuilder`.

File: operate/cli.py [hunk 2]
```diff
@@ -663,6 +666,98 @@ async def _stop_service_onchain(request: Request) -> JSONResponse:
...
+                chain_id=service.home_chain_id,
+                is_withdrawing=True,
+            )
+
+            service_manager.drain_service(
+                hash=request.path_params["service"],
+                withdrawal_address=withdrawal_address,
+            )
+
+            # drain OLAS from the master safe
```
Comment (jmoreira-valory) [inline operate/cli.py:702]: What if the Safe is needed for some other service?

File: operate/cli.py [hunk 3]
```diff
@@ -663,6 +666,98 @@ async def _stop_service_onchain(request: Request) -> JSONResponse:
...
+                    f"Draining {balance} OLAS out of master safe: {master_wallet.safe}"
+                )
+                transfer_erc20_from_safe(
+                    ledger_api=ledger_api,
+                    crypto=master_wallet.crypto,
+                    safe=t.cast(str, master_wallet.safe),
+                    token=OLAS[ledger_config.chain],
+                    to=withdrawal_address,
+                    amount=balance,
+                )
```
Comment (jmoreira-valory) [inline operate/cli.py:729]: Refactor in another PR: This should be done through the wallet/wallet manager, not here on the cli. Consider this for refactoring in another PR, or in the multi-agent branch.

File: operate/cli.py [hunk 4]
```diff
@@ -663,6 +666,98 @@ async def _stop_service_onchain(request: Request) -> JSONResponse:
...
+
+            # drain xDAI from the master safe
```
Comment (jmoreira-valory) [inline operate/cli.py:731]: Same comments apply.

File: operate/services/manage.py [hunk 5]
```diff
@@ -1332,6 +1348,80 @@ def fund_service(  # pylint: disable=too-many-arguments,too-many-locals
...
+            )
+            transfer_from_safe(
+                ledger_api=ledger_api,
+                crypto=ethereum_crypto,
+                safe=chain_data.multisig,
+                to=withdrawal_address,
+                amount=balance,
+            )
+
+        # drain xDAI from service signer key
```
Comment (jmoreira-valory) [inline operate/services/manage.py:1416]: Drain the agent wallet?

**Review by jmoreira-valory — APPROVED** [review]

File: operate/cli.py (hunk 2 above)
Comment (OjusWiZard) [reply operate/cli.py:702]: With the current design, it will need to be refilled

File: operate/services/manage.py [hunk 6]
```diff
@@ -1332,6 +1348,80 @@ def fund_service(  # pylint: disable=too-many-arguments,too-many-locals
...
+        chain_data = chain_config.chain_data
+        wallet_manager = self.wallet_manager.load(ledger_config.type)
+        ledger_api = wallet_manager.ledger_api(
+            chain_type=ledger_config.chain, rpc=ledger_config.rpc
+        )
+        ethereum_crypto = EthereumCrypto(
+            private_key_path=service.path / "deployment" / "ethereum_private_key.txt",
+        )
+
+        # drain OLAS and wxDAI from service safe
```
Comment (jmoreira-valory) [inline operate/services/manage.py:1370]: Possible more tokens need to be considered, e.g., USDC for Optimus

File: operate/cli.py (hunk 4 above)
Comment (OjusWiZard) [reply operate/cli.py:731]: With the current design, it will need to be refilled

File: operate/services/manage.py (hunk 5 above)
Comment (OjusWiZard) [reply operate/services/manage.py:1416]: yeah right

File: operate/services/manage.py (hunk 6 above)
Comment (OjusWiZard) [reply operate/services/manage.py:1370]: then, we much specify these tokens somewhere right? Perhaps in the service template?

**Pull request merged**